sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/CFBackedPageAsyncStream.java

```diff
@@ -0,0 +1,94 @@
+
+import java9.util.concurrent.CompletableFuture;
+import java9.util.concurrent.CompletionException;
+
+public final class CFBackedPageAsyncStream<T>
+    implements AsyncStream<PagedResponse<T>> {
+    private final Function<String, CompletableFuture<PagedResponse<T>>> pageRetriever;
+    private final Predicate<String> continuationPredicate;
+    private final String startPageId;
+    private final ClientLogger logger;
+
+    public CFBackedPageAsyncStream(Function<String, CompletableFuture<PagedResponse<T>>> pageRetriever,
+                                   Predicate<String> continuationPredicate,
+                                   String startPageId,
+                                   ClientLogger logger) {
+        this.pageRetriever = pageRetriever;
+        this.continuationPredicate = continuationPredicate;
+        this.startPageId = startPageId;
+        this.logger = logger;
+    }
+
+    @Override
+    public CancellationToken forEach(AsyncStreamHandler<PagedResponse<T>> handler) {
+        final CancellationToken token = new CancellationToken();
+        handler.onInit(token);
+        if (token.isCancellationRequested()) {
+            handler.onError(new CancellationException());
+            return token;
+        }
+
+        final CompletableFuture<Void> completableFuture = this.enumeratePages(startPageId, token, handler);
+        token.registerOnCancel(() -> {
+            completableFuture.cancel(true);
+        });
+        completableFuture.whenCompleteAsync((ignored, throwable) -> {
+            if (throwable != null) {
+                if (throwable instanceof CompletionException && throwable.getCause() != null) {
+                    // unwrap CF's CompletionException.
+                    handler.onError(throwable.getCause());
+                } else {
+                    handler.onError(throwable);
+                }
+            } else {
+                handler.onComplete();
+            }
+        });
+        return token;
+    }
+
+    private CompletableFuture<Void> enumeratePages(String pageId,
+                                                   CancellationToken token,
+                                                   AsyncStreamHandler<PagedResponse<T>> handler) {
+        return this.pageRetriever.call(pageId).handleAsync((pagedResponse, throwable) -> {
+            if (throwable != null) {
+                if (throwable instanceof RuntimeException) {
+                    // avoid double-wrapping for already unchecked exception
+                    throw logger.logExceptionAsError((RuntimeException) throwable);
+                } else {
+                    // wrap checked exception in a unchecked runtime exception
+                    throw logger.logExceptionAsError(new RuntimeException(throwable));
+                }
+            }
+
+            handler.onNext(pagedResponse);
+
+            final String nextPageId = pagedResponse.getContinuationToken();
```

Comment:
I see some variables named as "nextPageId" or "pageId" are actually refer to a "nextLink". Is it better to rename them as "nextLink"? Because "nextLink" is not an ID of the page. What do you think?

Reply:
This is a good question(I spent a lot of time thinking about it while designing); few overlapping use-cases make it hard to align on a good name. For example, android pagination libraries often use the name page identifier; for some azure services, it can be only a token, not a link (as in URL), and for some services, it's an HTTP/S URL.

collection::getPage(pageId) vs collection::getPage(token) vs collection::getPage(nextLink)

Out of various options, we felt that pageId suite better for the above API, as it is more common to be consumed from android pagination code user writes.

And PagedResponse::continuationToken(), covers nextLink if we treat it as a opaque string. 

But I agree that this is still not aligning perfectly. (Something we also want to observe in user study).